File: models/p8writer.lua
local M = {}

function hex_digit(digit)
  if digit < 10 then
      return "" .. digit
  else
      digit = math.min(digit, 15)
      return string.char(87 + digit)
  end
end

function byteToHex(byte, reverse)
  -- Convert negative numbers
  if byte < 0 then
    byte = 256 + byte
  end

  local lowNibble = hex_digit(byte % 16)
  local highNibble = hex_digit(math.floor(byte / 16))

  if reverse then
    return lowNibble .. highNibble
  else
    return highNibble .. lowNibble
  end
end

function M.writeP8(filename, bytes)
  local p8 = io.open(filename, "wb")

  p8:write("pico-8 cartridge // http://www.pico-8.com\n")
  p8:write("version 18\n")
  p8:write("__lua__\n\n")
  p8:write("#include main.lua\n")
  p8:write("#include decompressor.lua\n")
  p8:write("#include triangle.lua\n")
  p8:write("#include 3d.lua\n")
  p8:write("#include misc.lua\n")
  p8:write("#include objects.lua\n")
  p8:write("----------------\n")
  p8:write("__gfx__\n")

  local lineCount = 0
  local writingGfx = true
  for i, byte in ipairs(bytes) do
    if i == 8192 then
      writingGfx = false
      write("__map__\n")
    end
    local hex = byteToHex(byte, writingGfx)
    p8:write(hex)

    lineCount = lineCount + 1
    if (lineCount == 64 and writingGfx) or (lineCount == 128 and not writingGfx) then
      p8:write("\n")
      lineCount = 0
    end

  end

  p8:close()
end

return M

File: buildsystem/p8writer.js
const fs = require("fs");

module.exports.writeP8 = function (p8path, luaCode, graphicsData) {
  let p8 = "pico-8 cartridge // http://www.pico-8.com\nversion 23\n";

  if (luaCode != null && luaCode.length > 0) {
    p8 += `__lua__\n\n${luaCode}`;
  }

  if (graphicsData != null && graphicsData.length > 0) {
    p8 += "\n__gfx__\n";
    let lineCount = 0;
    let writingGfx = true;

    for (let c = 0; c < graphicsData.length; c++) {
      if (c == 8192) {
        writingGfx = false;
        p8 += "__map__\n";
      }
      let byte = graphicsData[c].toString(16);
      if (byte.length == 1) byte = `0${byte}`;
      if (writingGfx) {
        byte = byte.split("").reverse().join("");
      }
      p8 += byte;
      lineCount++;
      if ((lineCount == 64 && writingGfx) || (lineCount == 128 && !writingGfx)) {
        p8 += "\n";
        lineCount = 0;
      }
    }
  }

  fs.writeFileSync(p8path, p8);
}
